src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const pipeline = new Pipeline(p.name);
+  pipeline.steps = [];
+  for (let i = 0; i < p.steps.length; i++) {
+    let step: PotentialStep | null = cloneDeep(p.steps[i]);
+    step = await evaluateStep(step, conditionalCache);
+    if (step) {
+      pipeline.steps.push(step);
+    }
+  }
+  return pipeline;
+}
+
+async function evaluateStep(
+  step: PotentialStep,
+  conditionalCache: Map<any, any>,
+): Promise<Step | null> {
+  if (step instanceof Conditional) {
+    if (await step.accept()) {
+      step = await getAndCacheDependency(conditionalCache, step);
+    } else {
+      return null;
+    }
+  }
+  const deps = step.dependencies;
+  const newDeps: Set<Step> = new Set();
+  for (const dep of deps) {
+    const newDep = await evaluateStep(dep, conditionalCache);
+    if (newDep) {
+      newDeps.add(newDep);
+    }
+  }
+  step.dependencies = newDeps;
+  return step;
+}
+
+export async function walk(p: Pipeline, mutator: Mutators): Promise<Pipeline> {
+  const nodeCache = new Map<any, any>();
+
+  async function walkPipeline(p: Pipeline): Promise<Pipeline> {
+    const newSteps = [];
+    let newPipeline = cloneDeep(p);
+    for (const step of newPipeline.steps) {
+      let newStep = cloneDeep(step);
+      const result = await walkStep(newStep);
+      if (result) {
+        newStep = result;
+        newSteps.push(newStep);
+      }
+    }
+    newPipeline.steps = newSteps;
+    if (mutator.pipelineFn) {
+      newPipeline = mutator.pipelineFn(newPipeline);
+    }
+    return newPipeline;
+  }
+
+  async function walkStep(step: PotentialStep): Promise<Step> {
+    if (step instanceof Conditional) {
+      throw new Error(
+        `encountered conditional during walk, please run evaluatePipeline`,
+      );
+    }
+
+    if (nodeCache.has(step.key)) {
```

Comment:
I used the step key here since stringifying the entire object was throwing circular dependency errors. the step key is meant to be unique in the entire pipeline, so I figured this was a decent workaround

Reply:
keys are optional, not sure we can use this one. Just an idea: if you were to reuse the `sortedSteps` function that gives you a list of evaluated steps, couldn't you just call that, then mutate each step in the list and then put that back in order? The mutation could even be a parameter to `async toList(` of the `Pipeline` class. That way we don't have to reimplement almost the same thing two different ways and ensure the mutation is done the same way as the serialization. I think it would also be acceptable to only allow mutation with explicit dependencies, so we don't need to think about wait steps.

Reply:
I think that would cause repeated transformation on a step if it was used as a dependency in multiple steps.

Reply:
Why can't you use the step itself as the key?

Reply:
Typescript was throwing circular dependency errors, I thought the key was a good enough substitute that I didn't look that much into it.

They key seems to be lazily generated: https://github.com/joscha/buildkite-graph/blob/master/src/base.ts#L35 and it should be consistent across dependencies and nodes since it's the same object. The only thing is we'd be enforcing every step to generate a key before the serializer.

If there's enough doubt I can turn my attention back to just using the step object